plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java

```diff
@@ -814,25 +821,30 @@ public Optional<Object> getInfo(ConnectorTableHandle tableHandle)
     @Override
     public List<SchemaTableName> listTables(ConnectorSession session, Optional<String> optionalSchemaName)
     {
-        ImmutableSet.Builder<SchemaTableName> tableNames = ImmutableSet.builder();
-        for (String schemaName : listSchemas(session, optionalSchemaName)) {
-            for (TableInfo tableInfo : metastore.getTables(schemaName)) {
-                tableNames.add(tableInfo.tableName());
-            }
-        }
-        return tableNames.build().asList();
+        return streamTables(session, optionalSchemaName)
+                .map(TableInfo::tableName)
+                .collect(toImmutableList());
     }
 
     @Override
     public Map<SchemaTableName, RelationType> getRelationTypes(ConnectorSession session, Optional<String> optionalSchemaName)
     {
-        ImmutableMap.Builder<SchemaTableName, RelationType> result = ImmutableMap.builder();
-        for (String schemaName : listSchemas(session, optionalSchemaName)) {
-            for (TableInfo tableInfo : metastore.getTables(schemaName)) {
-                result.put(tableInfo.tableName(), tableInfo.extendedRelationType().toRelationType());
-            }
+        return streamTables(session, optionalSchemaName)
+                .collect(toImmutableMap(TableInfo::tableName, tableInfo -> tableInfo.extendedRelationType().toRelationType(), (ignore, second) -> second));
+    }
+
+    private Stream<TableInfo> streamTables(ConnectorSession session, Optional<String> optionalSchemaName)
+    {
+        List<Callable<List<TableInfo>>> tasks = listSchemas(session, optionalSchemaName).stream()
+                .map(schemaName -> (Callable<List<TableInfo>>) () -> metastore.getTables(schemaName))
+                .collect(toImmutableList());
+        try {
+            return processWithAdditionalThreads(tasks, metadataFetchingExecutor).stream()
```

Comment:
do query cancellations end up propagating properly? IIRC we had issues in past (related to metadata cache) where running a load and then cancelling the query didn't cancel ongoing loads and hence blocked all future metadata calls since executor was busy.

Can we test this (manually is fine too)?